Add tests for functions containing `utils::menu`

yangsophieee opened the pull request:
Relating to this issue (https://github.com/traitecoevo/traits.build/issues/41):
I've modified `metadata_add_locations` and implemented a test structure in `test-setup.R` for this function, based on [this thread](https://stackoverflow.com/a/65935322). 

I added an ifelse statement to `metadata_add_locations`, where the logic is that if `metadata_add_locations` is NOT run inside a `test_that` chunk, user input will be requested via `utils::menu` as normal. If it's run inside a `test_that` chunk then the function will simply return `names(location_data)`. Then in test-setup.R it tests if `metadata_add_locations` returns `names(location_data)`. I'm not sure what else to test, because right now it seems a bit pointless to test whether `names(location_data)` is what we expect. It doesn't test whether all the user input leads to the correct outcome. 

Within the else statement you could manually prescribe the first and second column of `location_data` to `location_name` and `keep` without user input and test the rest of the function that way. This feels clunky. 

It also feels wrong to modify the body of `metadata_add_locations` specifically for accommodating tests. 

I tried using the `mockr` package as suggested [here](https://stackoverflow.com/a/65740756) but couldn't get it to work. Another option is to replace `utils::menu` with `readLines` which seems quite promising (https://debruine.github.io/post/interactive-test/). Would you be happy with replacing `menu` with `readLines`?

Let me know what you think is the best option and I'll add similar modifications to `metadata_add_traits`, `metadata_add_contexts` and `metadata_create_template`.

dfalster commented on R/setup.R:
```diff
@@ -284,6 +291,8 @@ metadata_add_locations <- function(dataset_id, location_data) {
 
   write_metadata_dataset(metadata, dataset_id)
+
+  return(invisible(metadata))
```
even though we saved to file, this will return the object silently, which enables us to test it

dfalster commented on tests/testthat/test-setup.R:
```diff
@@ -102,6 +102,26 @@ test_that("metadata_add_source_bibtex is working", {
+      )
+    )
+  )
+  expect_equal(names(x$locations) , locations$site_name)
```
Now we can actually tests if the object was modified

yangsophieee commented on tests/testthat/functions.R:
```diff
@@ -142,3 +142,16 @@ test_structure <- function(
 }
+
+## a helper function to determine if this is being run as part of a test
+is_testing_env <- function() {
```
This function isn't being used in any code though is it?

dfalster replied:
No, it was, but then I switched strategy to providing inputs and using that for if statement. I N left it in FYI

yangsophieee commented:
Hi @dfalster, thank you so much! I completely agree that my testing wasn't actually testing the function but I wasn't sure how best to do it. Your code looks very elegant; thanks for lending your expertise 😄. Did you want me to remove the `is_testing_env` function? Happy to merge this in now.

dfalster reviewed (changes requested):
Can you add more tests on output, then replicate this strategy for other functions requiring user input?

yangsophieee commented:
I've added optional arguments for simulated user inputs for `metadata_add_locations`, `metadata_add_traits`, `metadata_add_contexts` and `metadata_create_template`. I've also revamped the console messages and added more testing for the functions. I also updated documentation of functions (all in the man folder; should I have done this on a separate branch?). 

There were a few functions that had no testing which I've since added tests for: `metadata_find_taxonomic_change`, `build_find_taxon`, `metadata_add_taxonomic_changes_list`, `metadata_add_substitutions_list`. Not sure how useful these last few functions are. The difference between `metadata_add_substitutions_table` and `metadata_add_substitutions_list` isn't large and the naming does not help to elucidate the differences. Maybe we could remove `metadata_add_substitutions_list` and rename `metadata_add_taxonomic_changes_list` to `metadata_add_taxonomic_changes_table`? I'm sure there are a lot of improvements that can still be made that future users will bring to our attention. For now I think it's good to merge and I can move on to fixing issues.

dfalster commented on tests/testthat.R:
```diff
@@ -1,4 +1,4 @@
-library(testthat)
+library(testthat) # Do you need this line if test_check has `testthat::` in front of it?
```
No, delete

dfalster commented on tests/testthat.R:
```diff
@@ -1,4 +1,4 @@
-library(testthat)
+library(testthat) # Do you need this line if test_check has `testthat::` in front of it?
 library(traits.build)
```
I think we can delete this too

dfalster commented on tests/testthat/test-utils.R:
```diff
@@ -1,15 +1,15 @@
 
-test_that("util_df_to_list is working", {
+test_that("`util_df_to_list` is working", {
   expect_type(util_df_to_list(ggplot2::mpg), "list")
```
Nice idea, but we don't want to rely on ggplot package for this, as it's not a core dependency and we should keep the footprint (list of dependencies as small as possible). Can we use one of the dfs in base R or dplyr (which is a core dependency)?

E.g. `iris` or `dplyr::band_instruments` or `dplyr::starters`?

dfalster commented on tests/testthat/test-utils.R:
```diff
@@ -1,15 +1,15 @@
+test_that("`util_list_to_df2` is working", {
   expect_equal(util_list_to_df2(NULL), NA)
   expect_equal(util_list_to_df2(NA), NA)
 
```
We should add a couple more tests, .e.g to show that the function produces output in desired format

`traits.build::util_df_to_list(dplyr::band_instruments)`

dfalster reviewed (approved):
Nice work @yangsophieee . Very thorough and needed. Before merging, please request a review from Lizzy, ask her to run everything as though she were inputting a new dataset. She doesn't need to review code.
